### DISCUSSION TITLE
Search Iterator as data option - complete?

### DISCUSSION THREAD

**Discussion by NicoN13** [anvilistas/tabulator#119] [Q&A]
Hi @s-cork 

I saw this post from early last year: https://github.com/s-cork/Tabulator/issues/19

Have you perhaps managed to implement search iterators as data options?

Thanks,

**Comment by s-cork** [comment]
If you read the readme you'll see app_table as an option. If not clear let me know what's missing.

**Reply by NicoN13** [reply]
Hi @s-cork 

Perhaps I misunderstood what you meant by the search iterator. I'm already using the app_table option by passing a client readable table. I thought the linked post had to do with passing the .search() result directly to the data property.

**Reply by s-cork** [reply]
When you use the `app_table` option, tabulator calls `.search()` on the app table you provide.
So it basically ends up being the same.

Passing `.search()` didn't end up being feasible. 
Mostly because - the JavaScript Tabulator implementation can't sort or query a search iterator.
JavaScript Tabulator needs all the data in order to implement sorting - i.e. a list of dicts.
And for a custom implementation I can't query a search iterator. I can only query an app table.


What problem would it solve being able to pass `.search()` over the `app_table` implementation?

**Reply by NicoN13** [reply]
Yeah that's how I understood it as well. I was hoping for improved speed by some sort of magic mumbo jumbo.

I return client_readable views from the server in less than half a second. But to populate the table takes long (5seconds). This is the case for all large data_tables. I've tried to improve it by using the initial_query including arguments like
q.page_size(25)
q.fetch_only(*relevant_columns)

Although these methods above do improve the speed ever so slightly, it's still not as quick as I was hoping.

My next approach was going to be to return the data from the server as a list of dicts, but only fetching say 25-50 rows at time and use my own pagination to control the data fetch and then passing this to the .data property.

Currently all my tables make use of the app_table option with {"progressive_load": "scroll"} which is great, but just too slow on the initial load.

**Reply by s-cork** [reply]
Happy to take a look at a clone link and see what we can do

**Reply by NicoN13** [reply]
I'll create a dummy app over the next day or so to show the challenge I'm having. Really appreciate the help by the way!

**Reply by NicoN13** [reply]
Hi @s-cork 

Haven't had time yet to create a dummy app with large enough tables and linked rows to showcase the problem. But I've been thinking, could it be the use of cell.get_table_row() in the column formatters that causes the slow down?

So let's say I have 5 tables.
Table1 has a linked column to Table2. All other columns in Table1 are numbers, strings, and so on. The only link is to Table2.
Table2 has linked columns to Table3, Table4 and Table5.

The client readable table is as such.
return app_tables.table1.client_readable()

Then my initial query is:
```
initial_query = {
    q.fetch_only(
        *table1_columns,
        table2_link=q.fetch_only(
            "product_name"
        )
    )
}
```

So I'm saying only fetch all table 1 columns and fetch the single column in the linked table to which I'm interested ("product_name"). However, in my column formatters I need to apply some formatting to the product_name in order to display it correctly. 
So in my column formatter, I say return cell.get_table_row()["table2_link"]["product_name"].upper() or any other conversion I might need to apply.

Could it be that calling cell.get_table_row() in effect fetches all columns and linked columns causing the slowdown?

**Reply by s-cork** [reply]
It shouldn't. It just return the row that was already fetched. If you start accessing columns that weren't included in fetch only it would slow down. But it doesn't sound like you are. 

Have a look at the browser console and see if you're getting a lot of unexpected rpc call logs. 

One rpc call represents one server fetch. If you're getting hundreds we know there's a problem. 

Ideally you'd get one per fetch of the page_size.

**Reply by NicoN13** [reply]
I can confirm I'm only getting one rpc call with the load, so it's not that.
